doc/cprover-manual/function-contracts.md

```diff
@@ -0,0 +1,502 @@
+# Function Contracts
+
+Every developer has the idea of an implementation in their heads. This is an
+implicit specification of their design, which others can not access. To formally
+describe these specifications, one may write contracts — an agreement between
+two parties specifying the obligations of the user and guarantees of the
+implementation. Contracts are beneficial to both sides of the agreement: the
+user can stop guessing their obligations and developers have the freedom to
+change/optimize the implementation, as long as it still obeys its contracts.
+
+Developers should ask themselves some questions to help them write function
+specifications:
+
+1. What are all possible values accepted by this function (i.e., full domain)?
+2. What are the caller responsibilities prior to invoking this function (i.e.,
+   preconditions)?
+3. What are the function implementation goals (i.e., postconditions)?
+4. Could the function be total (i.e., handle every possible input)?
+5. What are the edge cases for this function (e.g, error handling)?
+6. What are all side effects of this function (e.g., updates to global
+   variables)? Is this function pure (i.e., no side effects)?
+7. Which arguments or global variables may be modified by this function (i.e.,
+   writable set)? 
+8. Are there any ambiguities in this operation?
+
+To help developers proper describe these specifications, CBMC offers three basic
+function-contract clauses: _requires_, _ensures_, and _assigns_. These clauses
+accept either symbols or Boolean predicates that formally describe the
+specification of a function, but they might need more expressive constructs to
```

Comment:
function.  the "but they might" doesn't seem necessary.  I would remove it and just say, "CBMC also provides a series of built in constructs..."

Reply:
Removed it.